File: python3/add_two_numbers.py
"""
Space   : O(1)
Time    : O(n)
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        p1, p2 = l1, l2
        v1, v2 = 0, 0
        
        mul = 1
        while p1:
            v1 = v1 + (p1.val * mul)
            mul *= 10
            p1 = p1.next
        
        mul = 1
        while p2:
            v2 = v2 + (p2.val * mul)
            mul *= 10
            p2 = p2.next
        
        tot = str(v1 + v2)
        ans = ListNode(0)
        point = ans
        
        for i in range(len(tot)-1, -1, -1):
            point.next = ListNode(int(tot[i]))
            point = point.next
        
        return ans.next

File: golang/add_two_numbers.go
package golang

// Space    : O(1)
// Time     : O(n)

import "golang/model"

 func addTwoNumbers(l1 *model.ListNode, l2 *model.ListNode) *model.ListNode {
    ans := &model.ListNode{
        Val: 0,
    }
    p := ans
    carry := 0

    for l1 != nil || l2 != nil {
        v1, v2 := 0, 0
        
        if l1 != nil {
            v1 = l1.Val
        }
        
        if l2 != nil {
            v2 = l2.Val
        }
        
        v := (v1 + v2 + carry) % 10
        carry = (v1 + v2 + carry) / 10
        
        p.Next = &model.ListNode{
            Val: v,
        }
        p = p.Next
        
        if l1 != nil {
            l1 = l1.Next
        }
        
        if l2 != nil {
            l2 = l2.Next
        }
    }
    
    if carry != 0 {
        p.Next = &model.ListNode{
            Val: carry,
        }
    }
    
    return ans.Next
}